agents/agent.py

```diff
@@ -20,6 +21,55 @@
     'agents.utils.logtools', 'logconfig.ini'
 )
 
+MAX_RETRIES = 3
+RETRY_DELAY = 5
+
+
+def log_error(msg):
+    with open('agent_log.txt', 'a') as f:
+        timestamp = time.strftime('%Y-%m-%d %H:%M:%S')
+        f.write('[ERROR] %s - %s\n' % (timestamp, msg))
+
+
+def post_data(url, data, max_retries=MAX_RETRIES, delay=RETRY_DELAY):
+    """
+    Sends a POST request with JSON data to the specified URL
+    with retry logic. Retries up to `max_retries` times with `delay`
+    seconds between attempts. Logs all attempts and failures.
+    """
+    headers = {'Content-Type': 'application/json'}
+    payload = json.dumps(data).encode('utf-8')
+
+    for attempt in range(1, max_retries + 1):
+        try:
+            log_error(
+                '[Attempt %d] Sending data to %s' % (attempt, url)
+            )
+            request = urllib2.Request(
+                url, data=payload, headers=headers
+            )
+            response = urllib2.urlopen(request)
+            result = response.read()
+            response.close()
+            log_error(
+                'Success on attempt %d: %s' % (attempt, result)
+            )
+            return result
+        except urllib2.URLError as e:
+            log_error(
+                'Attempt %d failed: %s' % (attempt, e)
+            )
+            if attempt < max_retries:
+                log_error(
+                    'Retrying in %d seconds...' % delay
+                )
+                time.sleep(delay)
+            else:
+                log_error(
+                    'All %d attempts failed. Data not sent. '
+                    'Last error: %s' % (max_retries, e)
+                )
+
```

Comment:
If all attempts fail, `post_data()` returns `None`, which might be risky if the caller expects a response. You could raise an exception:
```
raise RuntimeError("POST failed after %d attempts" % max_retries)
```
Or explicitly return an error message or flag.